contracts/CommunityTokenDeployer.sol

```diff
@@ -0,0 +1,140 @@
+    {
+        if (
+            bytes(_ownerToken.name).length == 0 || bytes(_ownerToken.symbol).length == 0
+                || bytes(_ownerToken.baseURI).length == 0 || bytes(_masterToken.name).length == 0
+                || bytes(_masterToken.symbol).length == 0 || bytes(_masterToken.baseURI).length == 0
+        ) {
+            revert CommunityTokenDeployer_InvalidTokenMetadata();
+        }
+
+        if (_signature.signer == address(0) || _signerPublicKey.length == 0) {
+            revert CommunityTokenDeployer_InvalidSignerKeyOrCommunityAddress();
+        }
+
+        if (deployments[_signature.signer] != address(0)) {
+            revert CommunityTokenDeployer_AlreadyDeployed();
+        }
+
+        if (_signature.deadline < block.timestamp) {
+            revert CommunityTokenDeployer_DeploymentSignatureExpired();
+        }
+
+        if (_signature.deployer != msg.sender) {
+            revert CommunityTokenDeployer_InvalidDeployerAddress();
+        }
+
+        if (!_verifySignature(_signature)) {
+            revert CommunityTokenDeployer_InvalidDeploymentSignature();
+        }
+
+        OwnerToken ownerToken = new OwnerToken(
+          _ownerToken.name, 
+          _ownerToken.symbol, 
+          _ownerToken.baseURI, 
+          msg.sender,
+         _signerPublicKey
+        );
+        deployments[_signature.signer] = address(ownerToken);
+
+        MasterToken masterToken = new MasterToken(
+          _masterToken.name, 
+          _masterToken.symbol, 
+          _masterToken.baseURI,
+          address(ownerToken)
+        );
+        emit Deploy(address(ownerToken), address(masterToken));
+
+        return (address(ownerToken), address(masterToken));
+    }
+
+    function deployCREATE2(
+        TokenConfig calldata _ownerToken,
+        TokenConfig calldata _masterToken,
+        bytes memory _signerPublicKey,
+        address _communityAddress
+    )
+        external
+    {
+        OwnerToken ownerToken = new OwnerToken{
+          salt: bytes32(keccak256(abi.encodePacked(_communityAddress)))
+        }(
+          _ownerToken.name, 
+          _ownerToken.symbol, 
+          _ownerToken.baseURI, 
+          msg.sender,
+         _signerPublicKey
+        );
+        MasterToken masterToken = new MasterToken{
+          salt: bytes32(keccak256(abi.encodePacked(_communityAddress)))
+        }(
+          _masterToken.name, 
+          _masterToken.symbol, 
+          _masterToken.baseURI,
+          address(ownerToken)
+        );
+        emit Deploy(address(ownerToken), address(masterToken));
+    }
+
+    function DOMAIN_SEPARATOR() external view returns (bytes32) {
+        return _domainSeparatorV4();
+    }
+
+    function _verifySignature(DeploymentSignature calldata signature) internal view returns (bool) {
+        bytes32 digest = _hashTypedDataV4(
+            keccak256(
+                abi.encode(DEPLOYMENT_SIGNATURE_TYPEHASH, signature.signer, signature.deployer, signature.deadline)
```

Comment:
Typically, we'd want to maintain and a put a `signer` nonce in here to protect against signature replay attacks.
However, we already ensure in earlier checks that only one contract can be deployed for a given community.
So if a signature was successfully used, even if it was used for a replay attack, the contract will throw an error.